smart_card_connector_app/src/chrome-api-provider.js

```diff
@@ -0,0 +1,160 @@
+ * Returns the result code for chrome.smartCardProviderPrivate API that
+ * corresponds to the given PCSC error.
+ * @param {!PcscApi.ERROR_CODE} errorCode
+ * @return {!chrome.smartCardProviderPrivate.ResultCode}
+ * @throws {Error} Will throw if unknown error code encountered.
+ */
+function convertErrorCodeToEnum(errorCode) {
+  switch (errorCode) {
+    case PcscApi.SCARD_F_INTERNAL_ERROR:
+      return chrome.smartCardProviderPrivate.ResultCode.INTERNAL_ERROR;
+    // TODO(vkovalova): add other error codes
+    default:
+      throw new Error('Unknown error code ' + errorCode + ' encountered');
+  }
+}
+
+/**
+ * This class provides chrome.smartCardProviderPrivate API with PCSC responses.
+ *
+ * On creation, it subscribes to chrome.smartCardProviderPrivate API events,
+ * which correspond to different PCSC requests. When an event if fired,
+ * ChromeApiProvider sends a request to Pcsc.ServerRequestHandler and waits for
+ * the response. When the response is received, it reports the result back using
+ * chrome.smartCardProviderPrivate API.
+ */
+GSC.ConnectorApp.ChromeApiProvider = class extends goog.Disposable {
+  /**
+   * @param {!goog.messaging.AbstractChannel} serverMessageChannel
+   * @param {!Pcsc.ReadinessTracker} serverReadinessTracker
+   */
+  constructor(serverMessageChannel, serverReadinessTracker) {
+    super();
+    this.serverRequester_ = new Pcsc.ServerRequestHandler(
+        serverMessageChannel, serverReadinessTracker, 'chrome');
+
+    this.subscribeToChromeApiEvents_();
+
+    this.serverRequester_.addOnDisposeCallback(this.disposeInternal.bind(this));
+  }
+
+  /** @override */
+  disposeInternal() {
+    this.unsubscribeFromChromeApiEvents_();
+    this.serverRequester_.dispose()
+    super.disposeInternal()
+  }
+
+  /**
+   * Checks if chrome.smartCardProviderPrivate is available. Logs if not.
+   * @returns {boolean}
+   * @private
+   */
+  checkChromeAPIAvailable_() {
+    if (chrome.smartCardProviderPrivate === undefined) {
+      goog.log.warning(
+          logger, 'chrome.smartCardProviderPrivate API is not available');
+      return false;
+    }
+    return true;
+  }
+
+  /**
+   * Subscribes to chrome.smartCardProviderPrivate events.
+   * @private
+   */
+  subscribeToChromeApiEvents_() {
+    if (!this.checkChromeAPIAvailable_) {
+      return;
+    }
+
+    chrome.smartCardProviderPrivate.onEstablishContextRequested.addListener(
+        this.establishContextListener_.bind(this));
+  }
+  /**
+   * Unsubscribes from chrome.smartCardProviderPrivate events.
+   * @private
+   */
+  unsubscribeFromChromeApiEvents_() {
+    if (!this.checkChromeAPIAvailable_) {
+      return;
+    }
+
+    chrome.smartCardProviderPrivate.onEstablishContextRequested.removeListener(
+        this.establishContextListener_.bind(this));
+  }
+
+  /**
+   * @param {number} requestId
+   * @private
+   */
+  async establishContextListener_(requestId) {
+    const callArguments = [PcscApi.SCARD_SCOPE_SYSTEM, null, null];
```

Comment:
Yep, it would be nicer to use functions. But I don't like that with PcscApi we would add some extra steps until the message gets to C++ library: those function calls will be transformed into json messages, that later get routed to ClientHandler and then parsed by ClientHandler into remote call messages.

Reply:
Ack